### PR TITLE
[MRG] refactor & clean up database loading around MultiIndex class

### LINKED ISSUE #810: Would a "Directory" Index be useful?
I've been thinking about the idea of a `DirectoryIndex` that would replace the explicit functionality of directory traversal by encapsulating it within a subclass of `Index`. It seems like a nice abstraction, and as we develop further filtering and signature selection mechanisms (e.g. chaining) on top of `Index` objects, it'd be nice to have them apply to directory traversal.

The most basic form of `DirectoryIndex` would be something that simply traversed the directory every time `signatures()` was called. We could also imagine adding one that loads the signatures once and/or caches them, but that might get too memory intensive.

Any thoughts?

### LINKED ISSUE #1077: refactor LoadSingleSignatures?
The `sourmash_args.LoadSingleSignatures` class used by `categorize` is kinda ugly. could be done along with #829.

### LINKED ISSUE #1376: clean up `sourmash_args.load_dbs_and_sigs(...)` with scaled/num selectors
As I'm going through `sourmash_args.load_dbs_and_sigs(...)` for #1373 and #1374, I'm seeing a good opportunity for some refactoring and simplification around providing scaled/num selectors per https://github.com/dib-lab/sourmash/issues/1072

A few thoughts -
* if we added scaled/num selectors we could eliminate the `filter_compatible_signatures` check on lists of signatures;
* I think that would also remove the `check_lca_db_is_compatible` and `check_tree_is_compatible` checks;
* this would collapse the entire function down to something ...much smaller, although we might not be able to remove the conditionals on database/siglist type unless we wanted to change the `notify` output to just talk about total signatures loaded.

### PR DESCRIPTION
**Pull request by ctb**
This PR builds on #1374 to make database loading logic more generic. Specifically,
- move `load_from_directory` and `load_from_file_list` to `MultiIndex` class
- refactor `sourmash_args._load_database` accordingly

This results in substantial cleanup to `sourmash_args`, which is nice!

This is an intermediate step on the way to plugins, https://github.com/dib-lab/sourmash/pull/1368 #1353.

This PR also:
* contains #1420, which introduces expanded signature selection and compatibility checking for Index classes.
* fixes a bug in directory traversal that did _not_ yield `.sig.gz` files appropriately
* removes `LoadSingleSignature` from `sourmash categorize` (https://github.com/dib-lab/sourmash/issues/1077)
* refactors `sourmash lca summarize`/`command_summarize.load_singletons_and_count` to use `MultiIndex.load_from_path(...)` and `MultiIndex.signatures_with_location(...)` (#810)

## notes

As currently written, there are some changes to exceptions that may not be acceptable for semantic versioning.
* `load_file_as_index` now raises `ValueError` instead of `OSError`. See `test_api.py`.
* `load_signatures` now raises `ValueError` instead of a general `Exception` for a parse error.

Fixes https://github.com/dib-lab/sourmash/issues/1077
Fixes https://github.com/dib-lab/sourmash/issues/810
Fixes #1376
Addresses #1072

## TODO:
- [x] docstrings!
- [x] maybe further refactor out SIGLIST specific stuff? (punted to https://github.com/dib-lab/sourmash/pull/1420)
- [x] consolidate compatibility checking etc. (punted to https://github.com/dib-lab/sourmash/pull/1420)
- [x] test new MultiIndex load stuff directly (punted to https://github.com/dib-lab/sourmash/issues/1408)
- [x] improve debugging and exception reporting; maybe avoid overly broad try/except blocks 🙄 ...
- [x] create a specific exception/exception hierarchy for collection loading functions? (punted to #1417)
- [x] add tests for LCA DB `isfile` code

## Checklist

- [ ] Is it mergeable?
- [ ] `make test` Did it pass the tests?
- [ ] `make coverage` Is the new code covered?
- [ ] Did it change the command-line interface? Only additions are allowed
  without a major version increment. Changing file formats also requires a
  major version number increment.
- [ ] Was a spellchecker run on the source code and documentation after
  changes were made?

### CODE REVIEW

**Comment by ctb** [comment]
@luizirber I think this is at a good resting point and is some nice clean up. I'd be happy for a review and a merge if you have time to look at it and agree; I can move the TODO items to new issues. Alternatively, I will continue to poke at it to resolve some of the TODO items in this PR.

**Comment by ctb** [comment]
Ready for review and merge @luizirber @bluegenes!

File: src/sourmash/index.py [hunk 1]
```diff
@@ -193,6 +195,11 @@ def signatures(self):
             for ss in idx.signatures():
                 yield ss
 
+    def signatures_with_location(self):
+        for idx, loc in zip(self.index_list, self.source_list):
+            for ss in idx.signatures():
+                yield ss, loc
+
```
Comment (bluegenes) [inline src/sourmash/index.py:250]: oooh, location!

File: src/sourmash/index.py [hunk 2]
```diff
@@ -203,6 +210,53 @@ def insert(self, *args):
...
+        for thisfile in traverse_find_sigs([pathname], yield_all_files=force):
+            try:
+                idx = LinearIndex.load(thisfile)
+                index_list.append(idx)
+                source_list.append(thisfile)
+            except (IOError, sourmash.exceptions.SourmashError):
+                if force:
+                    continue    # ignore error
+                else:
+                    raise       # contine past error!
```
Comment (bluegenes) [inline src/sourmash/index.py]:
```suggestion
                    raise       # continue past error!
```

Comment (ctb) [reply src/sourmash/index.py]: fixed in 6d6eb42b

File: src/sourmash/index.py (hunk 1 above)
Comment (ctb) [reply src/sourmash/index.py:250]: yep :)

**Comment by ctb** [comment]
@bluegenes sorry I missed that you were reviewing this along with #1420 :). I resolved the issues you raised, let me know if there's more!

**Review by bluegenes — APPROVED** [review]

**Comment by ctb** [comment]
🎉

**Pull request merged**